membership.md

```diff
@@ -0,0 +1,83 @@
+
+This diagram, of which I'm quite proud, shows the possible transitions:
+```
++-----------------------------------------------------------------------------+
+|                                                                             |
+|                      Start ->  +--------+                                   |
+|            ,------<------------|        |                                   |
+|           /                    | absent |                                   |
+|          /       RemovePeer--> |        | <---RemovePeer                    |
+|         /            |         +--------+               \                   |
+|        /             |            |                      \                  |
+|   AddNonvoter        |         AddVoter                   \                 |
+|       |       ,->---' `--<-.      |                        \                |
+|       v      /              \     v                         \               |
+|  +----------+                +----------+                    +----------+   |
+|  |          | ---AddVoter--> |          | -log caught up --> |          |   |
+|  | nonvoter |                | staging  |                    |  voter   |   |
+|  |          | <-DemoteVoter- |          |                 ,- |          |   |
+|  +----------+         \      +----------+                /   +----------+   |
+|                        \                                /                   |
+|                         `--------------<---------------'                    |
+|                                                                             |
++-----------------------------------------------------------------------------+
+```
+
+While these operations aren't quite symmetric, we think they're a good set to capture
+the possible intent of the user. For example, if I want to make sure a server doesn't have a vote, but the server isn't part of the configuration at all, it probably shouldn't be added as a nonvoting server.
+
+Each of these application-level operations will be interpreted by the leader and, if it has an effect, will cause the leader to write a new configuration entry to its log. Which particular application-level operation caused the log entry to be written need not be part of the log entry.
+
+## Code implications
+
+This is a non-exhaustive list, but we came up with a few things:
+- Remove the PeerStore: the `peers.json` file introduces the possibility of getting out of sync with the log and snapshot, and it's hard to maintain this atomically as the log changes. It's not clear whether it's meant to track the committed or latest configuration, either.
+- Servers will have to search their snapshot and log to find the committed configuration and the latest configuration on startup.
+- Bootstrap will no longer use `peers.json` but should initialize the log or snapshot with an application-provided configuration entry.
```

Comment:
This has a little detail on how it's used: https://www.consul.io/docs/guides/outage.html. I think the typical use cases are these:

1. Loss of quorum in a multiple server cluster where you can't bring up new servers with the same IPs as the lost servers. You can stop the remaining servers, edit the peers.json to remove the permanently failed servers, and then restart them to try to regain quorum, and then add new servers into the mix.

2. Sync mis-match between raft peer state and cluster state. If there's a leadership transition at the time the event comes in to remove a server from the Consul cluster, there have been cases where there's a failed server in the Raft peers that's no longer in the Consul cluster, so there's currently no mechanism to remove the server from the Raft peers. This can likely be handled with some better operator commands, or a more robust syncing between these two sources of state, but the current fix for this is to stop all the servers and remove the orphaned server from the peers.json file.

The peers.json file is kind of a good escape hatch for situations where the cluster gets into a state where it can't make changes on its own, so a human operator needs to figure out what to do. Unfortunately, the way it is currently set up, it doesn't always take precedence (things in the Raft log can override what the operator wants), so I think it would be much better if we could have the cluster get into a safe state (or have a way for an operator to put the cluster into a safe state), and let the operator promote a specific server to be leader to get quorum of 1, and then promote servers back into voting status once things are stable.

It seems super clean if the cluster is in an outage state to pick the node you want to be a leader, tell it to force all other nodes into nonvoter status, have it elect itself, and then have it convert them all to voting (via staging and the stability check).